api/providers/tasks.py

```diff
@@ -0,0 +1,539 @@
+        upload.delete()
+        sentry.log_exception()
+        message = 'Bulk upload preparation failure: failed to create the rows.'
+        return handle_internal_error(initiator=initiator, provider=provider, message=message)
+    logger.info('[{}] rows successfully prepared.'.format(len(created_objects)))
+
+    logger.info('Updating job state ...')
+    upload.state = JobState.INITIALIZED
+    try:
+        upload.save()
+    except ValidationError:
+        upload.delete()
+        sentry.log_exception()
+        message = 'Bulk upload preparation failure: job state update failed'
+        return handle_internal_error(initiator=initiator, provider=provider, message=message)
+    logger.info('Job state updated')
+    logger.info('Bulk upload preparation finished: [upload={}, provider={}, schema={}, '
+                'initiator={}]'.format(upload.id, upload.provider._id, upload.schema._id, upload.initiator._id))
+
+
+@celery_app.task(name='api.providers.tasks.monitor_registration_bulk_upload_jobs')
+def monitor_registration_bulk_upload_jobs(dry_run=True):
+
+    logger.info('Checking registration bulk upload jobs ...')
+    bulk_uploads = RegistrationBulkUploadJob.objects.filter(state=JobState.INITIALIZED)
+    logger.info('[{}] pending jobs found.'.format(len(bulk_uploads)))
+
+    for upload in bulk_uploads:
+        logger.info('Picked up job [upload={}, hash={}]'.format(upload.id, upload.payload_hash))
+        upload.state = JobState.PICKED_UP
+        bulk_create_registrations.delay(upload.id, dry_run=dry_run)
+        if not dry_run:
+            upload.save()
+    if dry_run:
+        logger.info('Dry run: bulk creation started in dry-run mode and job state was not updated')
+    logger.info('[{}] jobs have been picked up and kicked off. This monitor task ends.'.format(len(bulk_uploads)))
+
+
+@celery_app.task()
+def bulk_create_registrations(upload_id, dry_run=True):
+
+    try:
+        upload = RegistrationBulkUploadJob.objects.get(id=upload_id)
+    except RegistrationBulkUploadJob.DoesNotExist:
+        # This error should not happen since this task is only called by `monitor_registration_bulk_upload_jobs`
+        sentry.log_exception()
+        message = 'Registration bulk upload job not found: [id={}]'.format(upload_id)
+        return handle_internal_error(initiator=None, provider=None, message=message)
+
+    # Retrieve bulk upload job
+    provider = upload.provider
+    auto_approval = upload.provider.bulk_upload_auto_approval
+    schema = upload.schema
+    initiator = upload.initiator
+    logger.info(
+        'Bulk creating draft registrations: [provider={}, schema={}, initiator={}, auto_approval={}]'.format(
+            provider._id,
+            schema._id,
+            initiator._id,
+            auto_approval,
+        ),
+    )
+
+    # Check registration rows and pick up them one by one to create draft registrations
+    registration_rows = RegistrationBulkUploadRow.objects.filter(upload__id=upload_id)
```

Comment:
Yep, the upload job can be picked up only once (if successful) by the monitor task and the creation task is only triggered by the monitor task.

However, I should have added the filter `is_picked_up=False` just in case the task is ran from the shell.